### PR TITLE
 SALTO-6522 new state file format to avoid line length limit

### PR DESCRIPTION
**Pull request by dantal4**
a new state file format, that would be backward compatible and deals with too-long JSON lines:
```
[] // deprecated
{} // deprecated
[] // deprecated
"" // deprecated
{"versions":["0.1.2"]}
{"updateDates": [{"dummy":"2023-01-09T15:57:59.322Z"}]}
{"elements":[serializedElement1, serializedElement2, ..., serializedElementN]} // each line is shorter than MAX_STRING_LENGTH
{"elements":[serializedElementN+1, serializedElementN+2, ...]}
... // more {"elements":[...]} rows, until we cover all the elements
{"pathIndices":[...]}
{"pathIndices":[...]}
{"pathIndices":[...]}
```

this way the data in the state file won’t be positional, so we’ll be able to chunk long JSON lines, and after parsing each line we’ll merge them back to one `elements` / `pathIndices` lists.

The new state dumping code can be disabled using the `SALTO_DUMP_STATE_WITH_LEGACY_FORMAT=1` env var.

---

_Additional context for reviewer_

This PR is rebased on https://github.com/salto-io/salto/pull/6461

TODO:
- [x] add tests

---
_Release Notes_: 
Core:
- New state file format to avoid line length limit

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/workspace/src/serializer/elements.ts [hunk 1]
```diff
@@ -160,11 +159,17 @@ function isSerializedClass(value: any): value is SerializedClass {
...
-export const serializeStream = async <T = Element>(
-  elements: T[],
-  referenceSerializerMode: 'replaceRefWithValue' | 'keepRef' = 'replaceRefWithValue',
-  storeStaticFile?: (file: StaticFile) => Promise<void>,
-): Promise<AsyncIterable<string>> => {
+export const serializeStream = async <T = Element>({
+  elements,
+  referenceSerializerMode = 'replaceRefWithValue',
+  storeStaticFile,
+  streamSerializer = lowerdashSerialize.getSerializedStream,
```
Comment (edenhassid) [inline packages/workspace/src/serializer/elements.ts:166]: why did you add this if you don't use it?

File: packages/workspace/src/serializer/elements.ts
Comment (edenhassid) [inline packages/workspace/src/serializer/elements.ts:1]: didn't you have tests to change here? for `serializeStream`

File: packages/workspace/src/workspace/path_index.ts
Comment (edenhassid) [inline packages/workspace/src/workspace/path_index.ts:1]: don't you need tests for this too?

**Review by edenhassid — COMMENTED** [review]
looks great!! 💯 🔥 added some small questions

File: packages/workspace/src/serializer/elements.ts
Comment (dantal4) [reply packages/workspace/src/serializer/elements.ts:1]: do you mean to call `serializeStream` with the non-default `streamSerializer` param?

Comment (edenhassid) [reply packages/workspace/src/serializer/elements.ts:1]: you changed the params to be an object, it feels weird that you didn't need to correct any test, especially when this function is exported

File: packages/workspace/src/workspace/path_index.ts
Comment (dantal4) [reply packages/workspace/src/workspace/path_index.ts:1]: `serializePathIndexByAccount` has no UTs....

File: packages/workspace/src/serializer/elements.ts
Comment (dantal4) [reply packages/workspace/src/serializer/elements.ts:1]: `serializeStream` has no UTs... it's tested only wrapped by the `serialize` function.

**Review by edenhassid — APPROVED** [review]
